lib/withOnyx.js

```diff
@@ -37,13 +37,25 @@ export default function (mapOnyxToState) {
                 // disconnected. It is a key value store with the format {[mapping.key]: connectionID}.
                 this.activeConnectionIDs = {};
 
+                const cachedState = _.reduce(mapOnyxToState, (resultObj, mapping, propertyName) => {
+                    const key = Str.result(mapping.key, props);
+                    const value = Onyx.tryGetCachedValue(key, mapping);
+
+                    if (value !== undefined) {
+                        // eslint-disable-next-line no-param-reassign
+                        resultObj[propertyName] = value;
+                    }
+
+                    return resultObj;
+                }, {});
+
+                // If we have all the data we need, then we can render the component immediately
+                cachedState.loading = _.size(cachedState) < requiredKeysForInit.length;
+
                 // Object holding the temporary initial state for the component while we load the various Onyx keys
-                this.tempState = {};
+                this.tempState = cachedState;
```

Comment:
I am not sure why do we set the tempState to the same cachedState as the normal state. I might be missing something but could you help me understand this? Should we update the comment above?

Reply:
The intention here is that `tempState` is still used in the same way as it was before. 

If we did not get all the values from the cache then the values are held in the `tempState` until all are "loaded". This way we avoid calling `setState()` each time a value comes back from storage. That was the original optimization added here and is a _good thing_ because otherwise you may have like several calls to setState() happening in a kind of burst before the first render (at least that's what we saw and why we changed how this works).

Though the question did make me wonder about something... if `this.tempState` and `this.state` are initialized with the same object reference then will we be mutating the state directly on this line:

https://github.com/Expensify/react-native-onyx/blob/80c6f001e27e798ed56cc4f3a39ee27c5efae375/lib/withOnyx.js#L117

Seems like that would be undesirable in the case where you did not initialize all the values in the constructor.

Reply:
Thanks for the write up. @hannojg what do you think of the concern Marc raised above?